File: frontend/src/services/feeService.ts
import api from './api'
import type { FeeRequest, FeeResponse } from '@/types/Fee'
import type { Page, PageParams } from '@/types/Page'
import { DEFAULT_PAGE_SIZE } from '@/types/Page'

export const feeService = {
  async findAll(params: PageParams = {}): Promise<Page<FeeResponse>> {
    const { data } = await api.get<Page<FeeResponse>>('/fees', {
      params: {
        search: params.search ?? '',
        page: params.page ?? 0,
        size: params.size ?? DEFAULT_PAGE_SIZE,
      },
    })
    return data
  },

  async findById(id: string): Promise<FeeResponse> {
    const { data } = await api.get<FeeResponse>(`/fees/${id}`)
    return data
  },

  async create(request: FeeRequest): Promise<FeeResponse> {
    const { data } = await api.post<FeeResponse>('/fees', request)
    return data
  },

  async update(id: string, request: FeeRequest): Promise<FeeResponse> {
    const { data } = await api.put<FeeResponse>(`/fees/${id}`, request)
    return data
  },

  async remove(id: string): Promise<void> {
    await api.delete(`/fees/${id}`)
  },
}

File: backend/src/main/java/com/MenuBank/MenuBank/fee/FeeService.java
package com.MenuBank.MenuBank.fee;

import com.MenuBank.MenuBank.common.MerchantContext;
import com.MenuBank.MenuBank.merchant.MerchantRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class FeeService {

    private final FeeRepository feeRepository;
    private final MerchantRepository merchantRepository;
    private final MerchantContext merchantContext;

    public FeeService(FeeRepository feeRepository,
                      MerchantRepository merchantRepository,
                      MerchantContext merchantContext) {
        this.feeRepository = feeRepository;
        this.merchantRepository = merchantRepository;
        this.merchantContext = merchantContext;
    }

    public FeeResponse create(FeeRequest request) {
        UUID merchantId = merchantContext.getMerchantId();

        if (feeRepository.existsByNameAndMerchantId(request.getName(), merchantId)) {
            throw new DuplicateFeeException("nome");
        }

        Fee fee = Fee.builder()
                .merchant(merchantRepository.getReferenceById(merchantId))
                .name(request.getName())
                .feeRate(request.getFeeRate())
                .build();

        Fee saved = feeRepository.save(fee);
        return toResponse(saved);
    }

    public FeeResponse findById(UUID id) {
        UUID merchantId = merchantContext.getMerchantId();
        Fee fee = feeRepository.findByIdAndMerchantId(id, merchantId)
                .orElseThrow(() -> new FeeNotFoundException(id));
        return toResponse(fee);
    }

    public Page<FeeResponse> findAll(String search, Pageable pageable) {
        UUID merchantId = merchantContext.getMerchantId();
        String term = search == null ? "" : search;
        return feeRepository.findAllByMerchantIdAndNameContainingIgnoreCase(merchantId, term, pageable)
                .map(this::toResponse);
    }

    public FeeResponse update(UUID id, FeeRequest request) {
        UUID merchantId = merchantContext.getMerchantId();
        Fee fee = feeRepository.findByIdAndMerchantId(id, merchantId)
                .orElseThrow(() -> new FeeNotFoundException(id));

        fee.setName(request.getName());
        fee.setFeeRate(request.getFeeRate());

        Fee saved = feeRepository.save(fee);
        return toResponse(saved);
    }

    public void delete(UUID id) {
        UUID merchantId = merchantContext.getMerchantId();
        if (!feeRepository.existsByIdAndMerchantId(id, merchantId)) {
            throw new FeeNotFoundException(id);
        }
        feeRepository.deleteByIdAndMerchantId(id, merchantId);
    }

    private FeeResponse toResponse(Fee fee) {
        return FeeResponse.builder()
                .id(fee.getId())
                .name(fee.getName())
                .feeRate(fee.getFeeRate())
                .build();
    }
}
